### PR TITLE
fix: subscriber attributes migration edge cases

### PR DESCRIPTION
**Pull request by aboedo**
This fixes edge cases in the subscriber attributes bug fixes: 

### case 1: 
- Alice sets attributes and syncs
- Alice is aliased to Bob
- Bob syncs attributes, but never deletes attributes for Alice because they've already been successfully synced. 

### case 2: 
- Attributes are set for more than 1 user before migrating to the new format
- Attributes are migrated to the new format only for the current user. 

### Solution: 
- Upon initializing, we look for subscriber attributes in the old format for all users and migrate them. 
- After this step, we look for subscriber attributes for all users but the current one, that are already synced. We delete them. 

### Missing: 
- more detailed tests for the edge cases described above 
- cleanup for readability

### CODE REVIEW

File: Purchases/Caching/RCDeviceCache.m [hunk 1]
```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
...
+    }
+    [self storeAttributesForAllUsers:attributesInNewFormat];
+}
+
+- (void)deleteSyncedSubscriberAttributesForOtherUsers {
+    NSDictionary<NSString *, NSDictionary *> *allStoredAttributes = self.storedAttributesForAllUsers;
+    NSMutableDictionary<NSString *, NSDictionary *> *filteredAttributes = [[NSMutableDictionary alloc] init];
+
+    NSString *currentAppUserID = self.cachedAppUserID;
+    NSParameterAssert(currentAppUserID);
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:325]: Is this to avoid a crash when `NSUserDefaults` is cleared?

File: Purchases/Caching/RCDeviceCache.m [hunk 2]
```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
...
-    NSMutableDictionary *currentAttributesForAppUserID = allSubscriberAttributes[appUserID]
-                                                         ?: @{}.mutableCopy;
-    NSMutableDictionary *mutableLegacyAttributes = legacyAttributes.mutableCopy;
-    [mutableLegacyAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+- (void)migrateSubscriberAttributes {
+    NSArray *appUserIDsWithLegacyAttributes = [self appUserIDsWithLegacyAttributes];
+    NSMutableDictionary *attributesInNewFormat = self.storedAttributesForAllUsers.mutableCopy;
+    for (NSString *appUserID in appUserIDsWithLegacyAttributes) {
+        NSDictionary *legacyAttributes = ([self valueForLegacySubscriberAttributes:appUserID] ?: @{}).mutableCopy;
+        NSMutableDictionary *currentAttributesForAppUserID = (legacyAttributes ?: @{}).mutableCopy;
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:308]: It looks like `legacyAttributes` can't be nil so I don't think `?: @{}`  is necessary in this line

File: Purchases/Caching/RCDeviceCache.m [hunk 3]
```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
...
-                                                    ?: @{}).mutableCopy;
-    NSMutableDictionary *currentAttributesForAppUserID = allSubscriberAttributes[appUserID]
-                                                         ?: @{}.mutableCopy;
-    NSMutableDictionary *mutableLegacyAttributes = legacyAttributes.mutableCopy;
-    [mutableLegacyAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+- (void)migrateSubscriberAttributes {
+    NSArray *appUserIDsWithLegacyAttributes = [self appUserIDsWithLegacyAttributes];
+    NSMutableDictionary *attributesInNewFormat = self.storedAttributesForAllUsers.mutableCopy;
+    for (NSString *appUserID in appUserIDsWithLegacyAttributes) {
+        NSDictionary *legacyAttributes = ([self valueForLegacySubscriberAttributes:appUserID] ?: @{}).mutableCopy;
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:307]: If this an `NSDictionary`, the `mutableCopy` is not needed

File: Purchases/Caching/RCDeviceCache.m [hunk 4]
```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
...
+        NSMutableDictionary *currentAttributesForAppUserID = (legacyAttributes ?: @{}).mutableCopy;
+        NSMutableDictionary *allAttributesForUser = legacyAttributes.mutableCopy;
+        [allAttributesForUser addEntriesFromDictionary:currentAttributesForAppUserID];
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:310]: I might be missing something here but:

If `currentAttributesForAppUserID` is a copy of `legacyAttributes` and  `allAttributesForUser` is also a copy of `legacyAttributes`, isn't this line redundant since `allAttributesForUser` is already equal to `currentAttributesForAppUserID`?

If this line is redundant, then `currentAttributesForAppUserID` and `allAttributesForUser` are the same, and therefore all this can be simplified to:

```
        NSDictionary *legacyAttributes = [self valueForLegacySubscriberAttributes:appUserID] ?: @{};
        attributesInNewFormat[appUserID] = legacyAttributes;
```

File: Purchases/Caching/RCDeviceCache.m (hunk 1 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:325]: it's for when `NSUserDefaults` is cleared, yes, but it's not really meant to prevent a crash - it'll crash anyway. Just slightly easier to debug if this happens. 
 it's a fairly common practice in Obj-c to `NSParamAssert` stuff that you want to guarantee is non-null, since no-ops and nullability issues can be annoying as hell to debug.

File: Purchases/Caching/RCDeviceCache.m (hunk 3 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:307]: good catch, I refactored and forgot to remove the mutableCopy part

File: Purchases/Caching/RCDeviceCache.m (hunk 2 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:308]: good catch

File: Purchases/Caching/RCDeviceCache.m (hunk 4 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:310]: yep. I got dizzy while writing this. I'll clean it up, thanks for catching

Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:310]: just realized what the mixup was here: 
I meant to make sure that when migrating, if there happen to be any attributes in the new format already, those are kept. 
so line 309 should have been 
```objective-c
NSMutableDictionary *currentAttributesForAppUserID = (self.storedAttributesForAllUsers ?: @{}).mutableCopy;
```
I'll update, try to clean up for readability and add a test for the specific case

Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:310]: although it should never happen, but still

**Comment by aboedo** [comment]
@vegaro ready for another pass

File: Purchases/Caching/RCDeviceCache.m (hunk 1 above)
Comment (vegaro) [reply Purchases/Caching/RCDeviceCache.m:326]: oh I see, got it! But the other PR that added the observer on NSUserDefaults will still work right?

File: Purchases/Caching/RCDeviceCache.m (hunk 4 above)
Comment (vegaro) [reply Purchases/Caching/RCDeviceCache.m]: > I meant to make sure that when migrating, if there happen to be any attributes in the new format already, those are kept.

Ok perfect then!

File: Purchases/Caching/RCDeviceCache.m (hunk 1 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:326]: yep. so the only way to reach this line with a nil appUserID _should_ be if you never had one in the first place.

Comment (vegaro) [reply Purchases/Caching/RCDeviceCache.m:326]: I noticed this is calling `self.cachedAppUserID` instead of passing it to `deleteSyncedSubscriberAttributesForOtherUsers` from the `RCIdentityManager`. Do you think we could have some sort of race condition if configure gets called twice? 🤔

**Comment by vegaro** [comment]
I left a question https://github.com/RevenueCat/purchases-ios/pull/233/files#r421843999

I think we are fine though

File: Purchases/Caching/RCDeviceCache.m (hunk 1 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:326]: that's a valid concern. 
however: this method is private, and only called within a synchronized block, so passing in the appUserID would have the same effect as reading it from disk, right? 
also, the configure method actually creates a new instance of purchases and its dependencies (including this one), so their variables are independent

Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:326]: merging for now, but let's keep the convo open
